Style Engine: add typography to backend

ramonjd opened the pull request:
Tracking Issue: https://github.com/WordPress/gutenberg/issues/38167

## What?

Following on from https://github.com/WordPress/gutenberg/pull/39446 that introduced spacing to the Style Engine backend, this PR takes on typography ⚔️ 

## Why?

To introduce the Style Engine incrementally at first, and not change any output for now.

## How?

1. Using an existing method to output inline styles `generate`
2. Creating a new method `get_classnames` to output classnames inc. slugs.

## Testing Instructions

For dynamic blocks (Site Title, Navigation, Post Title, Post Date for example), check that the typography is output correctly on the frontend.

I've been testing so far with the Site Title block. To test with all controls you can update that [block's JSON](https://github.com/WordPress/gutenberg/blob/eaa2e0a91a8ccac43fa2e39f60c37713bff165d7/packages/block-library/src/site-title/block.json) supports with the following:

```json
		"typography": {
			"fontSize": true,
			"lineHeight": true,
			"__experimentalFontFamily": true,
			"__experimentalTextTransform": true,
			"__experimentalFontStyle": true,
			"__experimentalFontWeight": true,
			"__experimentalLetterSpacing": true,
			"__experimentalTextDecoration": true,
			"__experimentalDefaultControls": {
				"fontSize": true,
				"lineHeight": true,
				"fontAppearance": true,
				"letterSpacing": true,
				"textTransform": true
			}
		}
```

Test:

- Preset color classnames and font sizez
- Inline styles

Some test code (using TwentyTwentyTwo theme)

```html
<!-- wp:site-title {"style":{"typography":{"lineHeight":1.7,"textTransform":"uppercase","letterSpacing":"113px","textDecoration":"line-through"},"color":{"background":"#c2c9d0"}},"fontSize":"large","fontFamily":"source-serif-pro"} /-->
```

Deprecated HTML with presets
```html
<!-- wp:navigation {"orientation":"horizontal","style":{"typography":{"textTransform":"var:preset|text-transform|lowercase","textDecoration":"var:preset|text-decoration|strikethrough","fontStyle":"var:preset|font-style|italic","fontFamily":"var:preset|font-family|someAwesomeFont-right-here","fontWeight":"var:preset|font-weight|100"}}} -->
<!-- wp:navigation-link {"label":"WordPress","url":"https://www.wordpress.org/"} /-->
<!-- /wp:navigation -->
```
## Screenshots or screencast <!-- if applicable -->

ramonjd committed:
Initial commit: add typography to style engine

ramonjd committed:
Adding whitespace to inline output

ramonjd committed:
Introduce spacing to inline styles
First stab at generating classnames based on slugs

ramonjd committed:
Removing generating styles based on an options path since we're not using it yet.
Adding a new public function to generate_classnames.
Updated tests.

ramonjd commented on lib/block-supports/typography.php:
```diff
@@ -92,117 +92,142 @@ function gutenberg_apply_typography_support( $block_type, $block_attributes ) {
-			}
-			$styles[] = sprintf( 'font-family: %s;', $font_family_custom );
+			// Before using classes, the value was serialized as a CSS Custom Property.
+			// We don't need to check for a preset when it lands in core.
```
> We don't need to check for a preset when it lands in core

**TODO**: check what this means

cc @ockham Do you know if this comment is still valid? Grabbed your name from https://github.com/WordPress/gutenberg/commit/1e17b5a5b4a3117e843dde015c128361c5ae059e#diff-2c54d3c8305590cf826f83511a9fd85d5b405470addd62edc3183ff9118177fbR115

Or would we need the code permanently for backwards compatibility?

ramonjd commented on packages/style-engine/class-wp-style-engine.php:
```diff
@@ -109,56 +192,49 @@ public function generate( $block_styles, $options = array() ) {
 		$rules = array();
 
-		// If a path to a specific block style is defined, only return rules for that style.
-		if ( isset( $options['path'] ) && is_array( $options['path'] ) ) {
```
We don't actually use these feature, so I'm removing it.

ramonjd commented on packages/style-engine/class-wp-style-engine.php:
```diff
@@ -109,56 +192,49 @@ public function generate( $block_styles, $options = array() ) {
 	 * @return array The class name for the added style.
 	 */
-	public static function get_css_box_rules( $style_value, $style_property ) {
+	public static function get_css_rules( $style_value, $style_property ) {
```
To match the clientside naming.

ramonjd commented on phpunit/block-supports/spacing-test.php:
```diff
@@ -152,7 +152,7 @@ function test_margin_with_individual_skipped_serialization_block_supports() {
 		$actual   = gutenberg_apply_spacing_support( $block_type, $block_atts );
 		$expected = array(
-			'style' => 'padding-top:1px;padding-right:2px;padding-bottom:3px;padding-left:4px;',
+			'style' => 'padding-top: 1px; padding-right: 2px; padding-bottom: 3px; padding-left: 4px;',
```
Updated to use spaces to improve readability.

andrewserong commented on lib/block-supports/typography.php:
```diff
@@ -92,117 +92,142 @@ function gutenberg_apply_typography_support( $block_type, $block_attributes ) {
 
-	if ( ! empty( $classes ) ) {
-		$attributes['class'] = implode( ' ', $classes );
+	$style_engine  = gutenberg_get_style_engine();
+	$inline_styles = $style_engine->generate(
+		array( 'typography' => $styles ),
+		array(
+			'inline' => true,
+		)
+	);
+
+	$classnames = $style_engine->get_classnames(
+		array( 'typography' => $classes ),
+		array(
+			'use_schema' => true,
+		)
+	);
```
Seeing these two calls side-by-side, I was wondering if we should consolidate these calls so that `get_classnames` becomes an implementation detail of `generate` within the style engine? Then, we could adjust the return signature of `generate` to be an array with keys for `classes` and a key for `inline_styles` (or whatever we want to name the keys).

By the time we get to refactoring the [elements block support](https://github.com/wordpress/gutenberg/blob/trunk/lib/block-supports/elements.php#L15) to use the style engine, we'll need to handle returning generated class names, so I wondered if it's worth factoring that in now?

ramonjd replied:
> Seeing these two calls side-by-side, I was wondering if we should consolidate these calls so that get_classnames becomes an implementation detail of generate within the style engine? 

I agree it's something we probably should do. 

I spent some time thinking about this before I landed on two methods, and I got stuck on the structure of the arguments signature more than the return. That is, how show we should tell the style engine to deal with the different values?

The `$block_styles` arg matches Gutenberg's style object and has a very particular structure, and I couldn't see how classname slugs would fit into that. 

We could always expand the options array. I did try that but wasn't satisfied with my approach as it appeared haphazard and unintuitive, and not related at all to the main argument `$block_styles`.

```php
generate( 
     $block_styles,
      [ classnames => $array_of_classnames ]
)
```

The first argument could be a collection of `[ $block_styles, $block_classnames ]` I suppose. 

Another thought I had was to extend the style value with metadata, which would mean the style engine had its own implementation of `$block_styles`. For example:

```
// Classname/slug
$style['color'] = [ 'classname' => 'crimson'  ]; 

// Regular style
$style['color'] = '#fff'; // or  [ 'value' => '#fff'  ];
``` 

The reason why I left it at two methods was so we could indeed have separate methods that we could ultimately call in a single `generate`.

So maybe the current `generate` should rather be named `get_styles` or whatever, and we create a new `generate` that calls `get_styles` or `get_classnames` accordingly.  🤷 

What do you `reckon` 😄 ?

andrewserong replied:
Oh, that _is_ an interesting challenge! Thanks for explaining the logic there, and apologies when I was reading this yesterday, I totally missed that the following lines are grabbing root level attributes, not values within the style object 😅:

```
$classes['fontSize'] = _wp_to_kebab_case( $block_attributes['fontSize'] );
$classes['fontFamily'] = _wp_to_kebab_case( $block_attributes['fontFamily'] );
```

I think we're now bumping into one of the tricky areas where there's a part of the feature here that we probably need to decide between "do we change how it works now, or later?".

As in the border PR (https://github.com/WordPress/gutenberg/pull/37770) I think there's been discussion in various places about whether these root level attributes will (long-term) better belong within the `style` object. Ideally, we'd have everything we need right there in the style object, whether that style object is coming from block attributes, or global styles.

However, we're not there yet!

So in the spirit of the approach we've been taking thus far, which is — let's see if we can refactor what's currently working (pretty much as-is) and then iteratively switch over to new approaches — I reckon we go in the direction you've been exploring 👍

> We could always expand the options array. I did try that but wasn't satisfied with my approach as it appeared haphazard and unintuitive, and not related at all to the main argument $block_styles.

You're right, it's tricky to come up with something that feels intuitive. I was kinda wondering if we could nearly have a value in the options array called `attributes` or `additionalAttributes` or something, which is a way to pass in attributes that are not part of the `style` object — but that could feel a bit awkward, too...

Basically, I was thinking of something like:

```
$style_engine->generate(
		array( 'typography' => $styles ),
		array(
			'inline' => true,
			'attributes' => array(
				'fontSize' => 'large', // just the raw attribute, let the style engine do the kebab-ing?
				'fontFamily' => 'papyrus'
			)
		)
	);
```

But I agree, it does feel awkward including those attributes separately in the list of options, so it might not be the right approach...

> Another thought I had was to extend the style value with metadata, which would mean the style engine had its own implementation of $block_styles

That's actually not such a bad idea 🤔😀  — if we think of this as an opportunity to come up with the structure of how we long-term hope to structure the style object so that it contains those root level attributes, if we go with this approach, it might be a good way to confidence check what that shape will eventually look like.

It could also mean we kinda have an implementation which is "if the values are all in the style object, we don't need to look for anything else, otherwise, migrate over those other attributes to the style object for backwards compat" — that way we might be able to handle the desired location in the style object without ever having to worry about migrating old attributes? Just a thought....

> The reason why I left it at two methods was so we could indeed have separate methods that we could ultimately call in a single generate.
> So maybe the current generate should rather be named get_styles or whatever, and we create a new generate that calls get_styles or get_classnames accordingly.

I really like this idea, it means we've got two inner methods that are doing different things, and we have `generate` as an abstraction that sits just above, so we can have a single entry point to generating styles 👍

andrewserong commented on packages/style-engine/class-wp-style-engine.php:
```diff
@@ -88,6 +132,45 @@ protected function get_block_style_css_rules( $style_value, $path ) {
+				}
+
+				$style_definition = _wp_array_get( static::BLOCK_STYLE_DEFINITIONS_METADATA, $style_definition['path'], null );
+				// If there is no stored schema we could generate classnames based on other properties such as the path or value or other prefix passed to options.
```
Nice — thanks for flagging where we'd likely extend this to add in the generated classnames behaviour we've been exploring separately 👍

andrewserong commented on packages/style-engine/class-wp-style-engine.php:
```diff
@@ -88,6 +132,45 @@ protected function get_block_style_css_rules( $style_value, $path ) {
+		}
+
+		$classnames = array();
+		$use_schema = isset( $options['use_schema'] ) ? $options['use_schema'] : false;
```
Does this need to be an option, or could we infer it based on the preference within `BLOCK_STYLE_DEFINITIONS_METADATA`? Because `$options` is set for the overall call to `generate` (or `get_classnames` in this case) I was wondering if `use_schema` is a little too broad, in that each property / support might need a different approach?

Or did you have in mind a use case where we really do want to be able to globally switch between the two approaches when calling the style engine? 🤔

ramonjd replied:
Good question.

We don't need this right now, true. I was just playing with an idea of being able to pass a custom schema to the style engine for styles that might not be defined in `BLOCK_STYLE_DEFINITIONS_METADATA `, like layout container classes `wp-%s`. 

`$options['use_schema']` is fairly generic, but it was supposed to mean "use the pattern/schema stored in the style engine if there is one".

I'm not really sure to be honest.

It's not needed for this PR, so I'll probably dump it until we have a real-world use case.

andrewserong commented on lib/block-supports/typography.php:
```diff
@@ -92,117 +92,142 @@ function gutenberg_apply_typography_support( $block_type, $block_attributes ) {
 
 		if ( $has_named_font_size ) {
-			$classes[] = sprintf( 'has-%s-font-size', _wp_to_kebab_case( $block_attributes['fontSize'] ) );
+			$classes['fontSize'] = _wp_to_kebab_case( $block_attributes['fontSize'] );
```
I was wondering if it'd be worth moving this call to convert the value to kebab case to the style engine, too, so that the caller doesn't have to worry about the detail of how the class name is ultimately constructed, and the style engine can be responsible for that?

What I have in mind, is whether it'd be feasible for the style engine to infer the class names based on the attributes passed in via the `style` object handed to `generate`, instead of having to call `get_classnames` separately with the `$classes` value 🤔

ramonjd replied:
> convert the value to kebab case to the style engine, too, so that the caller doesn't have to worry about the detail of how the class name is ultimately constructed

It probably should, but would it need to if it's inferring classnames and building them "in house" based on attributes alone? 

The tricky part of the puzzle here is building classnames based on existing patterns. For `typography` only font size received the kebab case treatment. That's the reason why I left it to the block support to pass the right slug, and kept the style engine "dumb" in that regard. 

I can't think of any reason not to do this - in fact I agree that we definitely should ensure that all classnames are consistent in that way - however if ever a block/block support _didn't_ want kebab case (an extreme edge case) and something else, we might have painted ourselves into a corner too early.

Once again, I'm not sure. Maybe we'll know as we add more supports to the mix.

andrewserong replied:
> Once again, I'm not sure. Maybe we'll know as we add more supports to the mix.

Good points! I think if we're not sure, then perhaps let's leave it in the block support so that we haven't painted ourselves into a corner, like you say. We could always move it over in a follow-up, if need be.

andrewserong reviewed (commented):
Great work @ramonjd, this is testing very nicely!

<img width="1274" alt="image" src="https://user-images.githubusercontent.com/14988353/162906766-abeec220-a852-4d83-8105-d084d8ce4406.png">

✅ Each of the Typography styles are rendered correctly on the front end
✅ The generated classnames appear to be rendered and injected correctly into the markup

I've left a few comments, I hope you don't mind — mostly poking at the idea of whether it'd be feasible for us to consolidate into a single call to `generate`, move the logic of inferring classnames from attribute values to the style engine, and return an array with keys for the inline styles and classes.

It probably doesn't make a huge difference to do that at this stage, but I wondered if it'd be easier to lay the groundwork for that now, in prep for other block supports that will likely use classnames (and generated / constructed or random id classnames) more verbosely or frequently (mostly thinking of elements and layout supports).

What do you think? I can imagine there's a good argument either way for keeping separate for now / versus consolidating now.

Happy to do more testing tomorrow!

ramonjd committed:
Perform kebab case normalization in the style engine.
Remove value_func for common style definitions since I'm pretty sure we'll be using the get_css_rules method for 80%++ of cases.

ramonjd committed:
Remove unused arg $options

ramonjd commented:
Alternative that includes typography, but takes into account classname generation for colors here:

- https://github.com/WordPress/gutenberg/pull/40332

ramonjd commented:
I think I'll close this in favour of 

- https://github.com/WordPress/gutenberg/pull/40332

Since that PR takes classnames/presets into account.
